src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java

```diff
@@ -622,6 +633,38 @@ public static double binomialCoefficient(final int n, final int k) {
         return Math.pow(10, log10BinomialCoefficient(n, k));
     }
 
+    /**
+     * Calculates the binomial coefficient as a whole number. Designed to prevent
+     * overflows even with very large numbers and to be very fast for small numbers.
+     *
+     * @param n total number of trials
+     * @param k number of successes
+     * @return the binomial coefficient
+     */
+    public static long exactBinomialCoefficient(final int n, final int k) {
```

Comment:
Perhaps we should document the reasons we need our own implementation and beef up correctness and tests, or just consider using an available implementation?

Just iterating up over n and k, it appears that this implementation begins to disagree with CombinatoricUtils.binomialCoefficient around n = 48:

```
n=48, k=21, MathUtils.exactBinomialCoefficient=22314239266529, CombinatoricsUtils.binomialCoefficient=22314239266528
n=48, k=22, MathUtils.exactBinomialCoefficient=27385657281649, CombinatoricsUtils.binomialCoefficient=27385657281648
n=48, k=26, MathUtils.exactBinomialCoefficient=27385657281649, CombinatoricsUtils.binomialCoefficient=27385657281648
n=48, k=27, MathUtils.exactBinomialCoefficient=22314239266529, CombinatoricsUtils.binomialCoefficient=22314239266528
n=49, k=18, MathUtils.exactBinomialCoefficient=11554258485617, CombinatoricsUtils.binomialCoefficient=11554258485616
n=49, k=19, MathUtils.exactBinomialCoefficient=18851684897585, CombinatoricsUtils.binomialCoefficient=18851684897584
n=49, k=20, MathUtils.exactBinomialCoefficient=28277527346377, CombinatoricsUtils.binomialCoefficient=28277527346376
n=49, k=21, MathUtils.exactBinomialCoefficient=39049918716426, CombinatoricsUtils.binomialCoefficient=39049918716424
n=49, k=22, MathUtils.exactBinomialCoefficient=49699896548179, CombinatoricsUtils.binomialCoefficient=49699896548176
n=49, k=23, MathUtils.exactBinomialCoefficient=58343356817427, CombinatoricsUtils.binomialCoefficient=58343356817424
n=49, k=24, MathUtils.exactBinomialCoefficient=63205303218877, CombinatoricsUtils.binomialCoefficient=63205303218876
n=49, k=25, MathUtils.exactBinomialCoefficient=63205303218877, CombinatoricsUtils.binomialCoefficient=63205303218876
...
```

I haven't carefully tracked down the sources of the discrepancies, but just from skimming the CombinatoricsUtils implementation, it looks like there's additional logic to account for possible overflow in intermediate quantities.

Aside from these discrepancies, there are differences in how the implementations handle overflow in the final, returned value. It appears this implementation can silently return incorrect values (e.g., n = 100, k =40 returns -1---and one might naively assume from the bounds in line 650 that we can safely handle this regime), whereas the CombinatoricsUtils implementation throws a MathArithmeticException.

Just skimming the rest of the code in this PR, it looks like this code is only intended to calculate indices into an array? So that might naturally delimit the reasonable use cases. Perhaps we should roll an encapsulated implementation that only accounts for these use cases here, if the CombinatoricsUtils implementation does not suffice?

How about for other use cases in the GATK that might want to make use of a general MathUtils implementation? In which regions of input space do we really need an exact/cheap/etc. calculation? Is it ever desired behavior to throw an exception, or do we prefer returning an approximate value? And so on.

I'm going to submit this review comment separately so you can go ahead and start thinking about it, if you like---I expect that it will take me a bit of time to become acquainted with both old/new versions of the rest of the code, anyway!